### DISCUSSION TITLE
How to use HSS:YMODEM to load QSPI

### DISCUSSION THREAD

**Discussion by JV73** [polarfire-soc/polarfire-soc-documentation#396] [HSS and U-Boot] [answered]
I found information on how to enable YMODEM support in the HSS.  Is there any documentation on how to use it ?  How do I create an HSS payload in the correct file format ?  Does it just push the file sent into the selected memory space starting at offset 0 ?

**Comment by rockoguido** [comment]
Hi.. can we assume that you've configured HSS to have YMODEM enabled, so the CLI now contains the YMODEM command?  If so then you use that command and you get the options that include YMODEM Receive, of course selecting that puts you in Rx mode.   The best option I've seen  for serving up a .bin file has been Teraterm via File->Transfer->Ymodem->Send

and.. think you are asking about the use of the HSS Payload Generator to generate a  .bin file yes it is something we don't document that well .    are you comfortable generating a YAML file that the hss payload generator requires.   The workflow also tends to be Linux centric

**Reply by rockoguido** [reply, marked as answer]
just adding on..  will itemize some steps

Ubuntu...
1.   sudo apt install libyaml-dev libelf-dev zlib1g-dev
2. git clone https://github.com/polarfire-soc/hart-software-services.git
3. cd hart-software-services/tools/hss-payload-generator
    contains Readme for building on Linux and Windows
Once you build the hss-payload-generator.   how to use...

I'll insert a YAML file here that I've used...

#
# HSS Payload Generator - buildroot configuration file
#

# First, we can optionally set a name for our image, otherwise one will be created dynamically
set-name: 'MPFS example programs for DDR'

#
# Next, we'll define the entry point addresses for each hart, as follows:
#
hart-entry-points: {u54_1: '0x80000000', u54_2: '0x80000000', u54_3: '0x00000000', u54_4: '0x00000000'}
#
# Finally, we'll define a payloads (source binary file) that will be placed at certain regions in memory
# The payload section is defined with the keyword payloads, and then a number of individual
# payload descriptors.
#
# Each payload has a name (path to its ELF/bin file), an owner-hart, and optionally 1-3 secondary-harts.
#
# Additionally, it has a privilege mode in which it will start execution.
#  * Valid privilege modes are PRV_M, PRV_S and PRV_U.
#
# Case only matters for the ELF path names, not the keywords.
#
payloads:
  ./mpfs-user-crypto-aes-cryptography.elf: {exec-addr: '0x80000000', owner-hart: u54_1, secondary-hart: u54_2, priv-mode: prv_m, skip-opensbi: true}


This is for a simple bare metal app case that uses the first 2 harts.

I hope this helps..
Dave

once your .elf file is ready it's just ..

hss-payload-generator -c xxx.yaml   output.bin

This bin file is now HSS ready

**Reply by AustinOwens** [reply]
@rockoguido I am using minicom for ymodem file transfers, but it doesn't seem like the file is able to transfer correctly.  I keep getting NAK messages like this:

```bash
Ymodem sectors/kbytes sent: 664/83k
Retry 0: Got 1b for sector ACK
Retry 0: NAK on sector                                           
Retry 0: Got 0d for sector ACK                                   
Retry 0: NAK on sector                                           
Retry 0: Got 68 for sector ACK         
Retry 0: NAK on sector                            
Retry 0: Got 20 for sector ACK
Retry 0: NAK on sector
Retry 0: Retry Count Exceeded
Transfer incomplete
```

Has anyone seen this before?

**Comment by rockoguido** [comment]
Hi..

Would probably need more information as to the use case.   We do support Ymodem via our MPFS_HAL_DDR_DEMO, essentially this is one of our bootloaders.
I have used this before on the Polarfire Icicle Kit and I used Ymodem via Teratem to send the file to the Icicle Kit.

Regards,

Dave Campanella
Principal Embedded Solutions Engineer / FAE
Microchip Technology Inc.
165 Technology Dr
Irvine, CA 92618
+1 925 989 0569 (mobile)
+1 760 304 0172 (office)
***@***.******@***.***>

For any Technical related concerns, kindly you can create web ticket at http://www.microchip.com/support
For Technical Training please go to Microchip University at https://secure.microchip.com/mu

There are several paths to take for Microchip support:
        * Self-paced learning – see the Developer Help Site<http://microchipdeveloper.com/>
        * Knowledge Base - here<http://www.microchip.com/support/hottopics.aspx>
        * Community Forums - help<http://www.microchip.com/forums/> with a lot of information from experts.

**Reply by AustinOwens** [reply]
The use case is the same as above, I want to pass in payload to flash to QSPI. The first step though is just getting the file over, which is number 3 in the YMODEM util:

![image](https://github.com/user-attachments/assets/9d0f4105-7d77-4ece-a084-a17fee9d42fe)

I compiled HSS with QSPI and YMODEM services, as well as modified the constraints file in Libero to ensure that they are being pushed to the correct header pins. I also had to disable watchdog from HSS because the file I am trying to transfer over takes longer than 30 seconds to transfer (it's the core-image-minimal-mtdutils-icicle-kit-es-nor.nor.mtdimg from the yocto build).

In minicom when/after the file transfers, I get those "NAK on sector" messages. 

Just to make sure, I don't need to use the same USB port on the icicle that "USBDMSC" would typically use, correct? My assumption is that the UART channel that YMODEM is using is the same one that HSS uses (channel 0).

**Comment by rockoguido** [comment]
Yes you are correct in the choice of the UART for Ymodem.   To be honest I’ve never used Ymodem from HSS, typically have loaded images via USB and the USBMSC command in HSS and then using dd from the Linux side and sending the image over the USB channel and when xfer is done using a ctrl-C to xfer to the  QSPI flash itself via HSS

Assuming you have the NAND or NOR flash enabled that we support out of the box.  Devices called out in this app note..

https://github.com/polarfire-soc/polarfire-soc-documentation/blob/master/how-to/booting-from-qspi.md

I assume as a test case you have the USBMSC scenario working?

Regards,




Dave Campanella
Principal Embedded Solutions Engineer / FAE
Microchip Technology Inc.
165 Technology Dr
Irvine, CA 92618
+1 925 989 0569 (mobile)
+1 760 304 0172 (office)
***@***.******@***.***>

For any Technical related concerns, kindly you can create web ticket at http://www.microchip.com/support
For Technical Training please go to Microchip University at https://secure.microchip.com/mu

There are several paths to take for Microchip support:
        * Self-paced learning – see the Developer Help Site<http://microchipdeveloper.com/>
        * Knowledge Base - here<http://www.microchip.com/support/hottopics.aspx>
        * Community Forums - help<http://www.microchip.com/forums/> with a lot of information from experts.

**Reply by AustinOwens** [reply]
I am pretty confident I can get USBMSC working. When I invoke the util from HSS, it says that it has initialized QSPI and its waiting for bytes to by written from host. I can't test this right now because the board is remote to me right now and there is no USB plugged into J16 (the usb near the SD card slot).

However, it's sort of a moot point...I will not have access to that USB and will only have UART access in a custom PCB spin. 

Thank you for your fast responses by the way, very much appreciated.

**Comment by rockoguido** [comment]
Ok understood more about your environment.    Would recommend opening a support ticket, see the link below.

If it’s an Icicle kit we can replicate pretty easily for sure…

Regards,

Dave Campanella
Principal Embedded Solutions Engineer / FAE
Microchip Technology Inc.
165 Technology Dr
Irvine, CA 92618
+1 925 989 0569 (mobile)
+1 760 304 0172 (office)
***@***.******@***.***>

For any Technical related concerns, kindly you can create web ticket at http://www.microchip.com/support
For Technical Training please go to Microchip University at https://secure.microchip.com/mu

There are several paths to take for Microchip support:
        * Self-paced learning – see the Developer Help Site<http://microchipdeveloper.com/>
        * Knowledge Base - here<http://www.microchip.com/support/hottopics.aspx>
        * Community Forums - help<http://www.microchip.com/forums/> with a lot of information from experts.

**Reply by AustinOwens** [reply]
Thank you for the help!
